### PR TITLE
BUG: Fix `arr.flat.index` for large arrays and big-endian machines

### LINKED ISSUE #19153: ``flatiter``  when big-endian does not update index
There is a difference in the operations of ``numpy.flatiter`` between little-endian versus big-endian machines. Specifically, the ``.index`` and ``.coords`` attributes update on the ``flatiter`` instances when little-endian, but not when big-endian.

CC: @pllim, @mhvk
See https://github.com/astropy/astropy/issues/11804 and https://github.com/astropy/astropy/pull/11806 for initial bug report and testing of the bug, respectively.

### Reproducing code example:

```python
import numpy as np
x = np.arange(6).reshape(3, 1, 2)
xf = x.flat
endindices = [(xf.index, xf.coords) for _ in xf][-2] # next() oversteps

assert endindices[0] == 5  # should be 5, will be 0 on big-endian
```

**This test will only fail on big-endian machines**

### Error message:

``AssertionError``  The value is 0, not 5.

### NumPy/Python version information:

Platform: Linux-5.4.0-1047-azure-s390x-with-debian-10.9
Numpy : 1.20.3
Python: 3.7.3

### PR DESCRIPTION
**Pull request by seberg**
To do that, I have a first commit to introduce a new `PyArray_PyIntFromIntp` inline function, just because those `#ifdef`'s everywhere are not very nice IMO.

Closes gh-19153

### CODE REVIEW

File: numpy/core/src/multiarray/iterators.c [hunk 1]
```diff
@@ -1410,31 +1415,13 @@ arraymultiter_dealloc(PyArrayMultiIterObject *multi)
 static PyObject *
 arraymultiter_size_get(PyArrayMultiIterObject *self)
 {
-#if NPY_SIZEOF_INTP <= NPY_SIZEOF_LONG
-    return PyLong_FromLong((long) self->size);
-#else
-    if (self->size < NPY_MAX_LONG) {
```
Comment (seberg) [inline numpy/core/src/multiarray/iterators.c:1416]: These must just leftovers from Python 2

**Comment by seberg** [comment]
Actually, this also needs to be used in at least all of `nditer_pywrap.c` to unbreak windows for large iterators...

**Review by mhvk — COMMENTED** [review]
That seems all good! But what exactly went wrong on bigendian?

**Comment by seberg** [comment]
The `INT` is wrong and has half size, so the actual intp, has the size of `<int_upper><int_lower>` on big endian and `<int_lower><int_upper>` on little, so you only get the all zero high bytes on big-endian.  (or something along those lines...)

**Comment by mhvk** [comment]
Failure unrelated (see #19285)

**Comment by mattip** [comment]
Thanks @seberg

**Comment by pllim** [comment]
Thank you, all!

**Pull request merged**